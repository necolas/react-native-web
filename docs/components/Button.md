# Button

A basic button component. Supports a minimal level of customization. You can
build your own custom button using `TouchableOpacity` or
`TouchableNativeFeedback`.

## Props

**accessibilityLabel**: ?string

Overrides the text that's read by a screen reader when the user interacts
with the element.

**color**: ?string

Background color of the button.

**disabled**: ?boolean

If `true`, disable all interactions for this element.

**onPress**: function

This function is called on press.

**testID**: ?string

Used to locate this view in end-to-end tests.

**title**: string

Text to display inside the button.

## Examples

```js
<Button
  accessibilityLabel="Learn more about this purple button"
  color="#841584"
  onPress={onPressLearnMore}
  title="Learn More"
/>
```
