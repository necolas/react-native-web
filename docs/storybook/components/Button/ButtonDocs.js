/* eslint-disable react/jsx-sort-props */

/**
 * @flow
 */

import React from 'react';
import PropColor from './examples/PropColor';
import PropDisabled from './examples/PropDisabled';
import PropOnPress from './examples/PropOnPress';
import UIExplorer, { AppText, Code, DocItem } from '../../ui-explorer';
import { storiesOf } from '@kadira/storybook';

const sections = [
  {
    title: 'Props',
    entries: [
      <DocItem
        name="accessibilityLabel"
        typeInfo="?string"
        description="Overrides the text that's read by a screen reader when the user interacts with the element."
      />,

      <DocItem
        name="color"
        typeInfo="?string"
        description="Background color of the button."
        example={{
          render: () => <PropColor />
        }}
      />,

      <DocItem
        name="disabled"
        typeInfo="?boolean"
        description="If `true`, disable all interactions for this element."
        example={{
          render: () => <PropDisabled />
        }}
      />,

      <DocItem
        name="onPress"
        typeInfo="function"
        description="This function is called on press."
        example={{
          render: () => <PropOnPress />
        }}
      />,

      <DocItem
        name="testID"
        typeInfo="?string"
        description="Used to locate this view in end-to-end tests."
      />,

      <DocItem name="title" typeInfo="string" description="Text to display inside the button." />
    ]
  }
];

storiesOf('Components', module).add('Button', () =>
  <UIExplorer
    description={[
      <AppText>
        A basic button component. Supports a minimal level of customization.
        You can build your own custom button using <Code>TouchableOpacity</Code>
        or <Code>TouchableNativeFeedback</Code>.
      </AppText>
    ]}
    sections={sections}
    title="Button"
    url="components/Button"
  />
);
