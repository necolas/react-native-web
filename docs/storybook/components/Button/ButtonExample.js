import React from 'react';
import UIExplorer from '../../UIExplorer';
import { action, storiesOf } from '@kadira/storybook';
import { Button, StyleSheet, View } from 'react-native';

const onButtonPress = action('Button has been pressed!');

const examples = [
  {
    title: 'Default',
    description:
      'The title and onPress handler are required. It is ' +
        'recommended to set "accessibilityLabel" to help make your app usable by ' +
        'everyone.',
    render() {
      return (
        <Button
          accessibilityLabel="See an informative alert"
          onPress={onButtonPress}
          title="Press me"
        />
      );
    }
  },
  {
    title: 'Custom colors',
    description:
      'Adjusts the color in a way that looks standard on each ' +
        'platform. On iOS, the color prop controls the color of the text. On ' +
        'Android, the color adjusts the background color of the button.',
    render() {
      return (
        <View>
          <Button color="#17BF63" onPress={onButtonPress} title="Press me" />
          <View style={styles.verticalDivider} />
          <Button color="#F45D22" onPress={onButtonPress} title="Press me" />
          <View style={styles.verticalDivider} />
          <Button color="#794BC4" onPress={onButtonPress} title="Press me" />
          <View style={styles.verticalDivider} />
          <Button color="#E0245E" onPress={onButtonPress} title="Press me" />
        </View>
      );
    }
  },
  {
    title: 'Disabled',
    description: 'All interactions for the component are disabled.',
    render() {
      return <Button disabled onPress={onButtonPress} title="Disabled button" />;
    }
  },
  {
    title: 'Fit to text layout',
    description: 'This layout strategy lets the title define the width of the button',
    render() {
      return (
        <View style={styles.horizontal}>
          <Button
            accessibilityLabel="This sounds great!"
            onPress={onButtonPress}
            title="This looks great!"
          />
          <View style={styles.horizontalDivider} />
          <Button
            accessibilityLabel="Ok, Great!"
            color="#841584"
            onPress={onButtonPress}
            title="Ok!"
          />
        </View>
      );
    }
  }
];

const styles = StyleSheet.create({
  horizontalDivider: {
    width: '0.6rem'
  },
  horizontal: {
    flexDirection: 'row'
  },
  verticalDivider: {
    height: '1.3125rem'
  }
});

storiesOf('Components', module).add('Button', () =>
  <UIExplorer
    description="A basic button component. Supports a minimal level of customization. You can build your own custom button using &quot;TouchableOpacity&quot; or &quot;TouchableNativeFeedback&quot;"
    examples={examples}
    title="Button"
    url="https://github.com/necolas/react-native-web/blob/master/docs/components/Button.md"
  />
);
