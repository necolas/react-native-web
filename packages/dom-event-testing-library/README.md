# `dom-event-testing-library`

A library for unit testing high-level interactions via simple pointer events, e.g., `pointerdown`, that produce realistic and complete DOM event sequences.

There are number of challenges involved in unit testing modules that work with DOM events.

1. Testing environments with and without support for the `PointerEvent` API.
2. Testing various user interaction modes including mouse, touch, and pen use.
3. Testing against the event sequences browsers actually produce (e.g., emulated touch and mouse events.)
4. Testing against the event properties DOM events include (i.e., more complete mock data)
5. Testing against "virtual" events produced by tools like screen-readers.

Writing unit tests to cover all these scenarios is tedious and error prone. This event testing library is designed to avoid these issues by allowing developers to more easily dispatch events in unit tests, and to more reliably test interactions while using an API based on `PointerEvent`.

## Example

```js
import {
  describeWithPointerEvent,
  testWithPointerType,
  clearPointers,
  createEventTarget,
  setPointerEvent,
} from 'dom-event-testing-library';

describeWithPointerEvent('useTap', hasPointerEvent => {
  beforeEach(() => {
    // basic PointerEvent mock
    setPointerEvent(hasPointerEvent);
  });

  afterEach(() => {
    // clear active pointers between test runs
    clearPointers();
  });

  // test all the pointer types supported by the environment
  testWithPointerType('pointer down', pointerType => {
    const ref = createRef(null);
    const onTapStart = jest.fn();

    // component to test
    function Component() {
      useTapEvents(ref, { onTapStart });
      return <div ref={ref} />
    }

    // render component
    act(() => {
      render(<Component />);
    });

    // create an event target
    const target = createEventTarget(ref.current);

    // dispatch high-level pointer event
    act(() => {
      target.pointerdown({ pointerType });
    });

    // assertion
    expect(onTapStart).toBeCalled();
  });
});
```

The example above tests the interaction in multiple scenarios. In each case, a realistic DOM event sequence–with complete mock events–is produced. When running in a mock environment without the `PointerEvent` API, the test runs for both `mouse` and `touch` pointer types. When `touch` is the pointer type it produces emulated mouse events. When running in a mock environment with the `PointerEvent` API, the test runs for `mouse`, `touch`, and `pen` pointer types.

It's important to cover all these scenarios because it's very easy to introduce bugs – e.g., double calling of callbacks – if not accounting for emulated mouse events, differences in target capturing between `touch` and `mouse` pointers, and the different semantics of `button` across event APIs.

Default values are provided for the expected native events properties. They can also be customized as needed in a test.

```js
target.pointerdown({
  button: 0,
  buttons: 1,
  pageX: 10,
  pageY: 10,
  pointerType,
  // NOTE: use x,y instead of clientX,clientY
  x: 10,
  y: 10
});
```

Tests that dispatch multiple pointer events will dispatch multi-touch native events on the target.

```js
// first pointer is active
target.pointerdown({pointerId: 1, pointerType});
// second pointer is active
target.pointerdown({pointerId: 2, pointerType});
```

## API

### Target and events

To create a new event target pass the DOM node to `createEventTarget(node)`. This target can then be used to dispatch event sequences and customize the event payload. The following are currently supported:

* `blur`
* `click`
* `contextmenu`
* `focus` (includes the complete sequence of focus-related events)
* `keydown`
* `keyup`
* `pointercancel`
* `pointerdown`
* `pointerhover` (moves when pointer is not down)
* `pointermove` (moves when pointer is down)
* `pointerover`
* `pointerout`
* `scroll`
* `select`
* `selectionchange`
* `tap` (equivalent to `pointerdown` followed by `pointerup`)
* `virtualclick`

The target also has `node` property equal to the node that was used to create the target, and a `setBoundClientRect({x,y,width,height})` method that can be used to mock the return value of `getBoundingClientRect`.

### Jest helpers

#### `describeWithPointerEvent`

This is just like `describe` but it will run the entire test suite twice, once in an environment with `PointerEvent` mocked and once without.

```js
describeWithPointerEvent('useTap', hasPointerEvent => {
  // test suite
});
```

#### `testWithPointerType`

The is just like `test` but it will run the test for every pointer type supported by the environment. When `PointerEvent` is mocked, the pointer types will be `mouse`, `touch`, and `pen`; otherwise the pointer types will be `mouse` and `touch`.

```js
testWithPointerType('pointer down', pointerType => {
  // test unit
});
```

### jsdom environment helpers

#### platform

Interactions that account for Windows / macOS differences can change the platform by calling `platform.set(value)`, where `value` can be either `'mac'` or `'windows'`. To retreive the current platform call `platform.get()`, and the clear it call `platform.clear()`.

#### hasPointerEvent / setPointerEvent

Interactions implemented using `PointerEvent` can create a basic mock for jsdom by calling `setPointerEvent(true)` (disable with `setPointerEvent(false)`), and check whether `PointerEvent` is available by calling `hasPointerEvent()`.
