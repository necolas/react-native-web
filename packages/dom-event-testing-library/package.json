{
  "private": true,
  "name": "dom-event-testing-library",
  "version": "0.0.0",
  "main": "index.js",
  "description": "Browser event sequences for unit tests",
  "author": "Nicolas Gallagher",
  "license": "MIT",
  "homepage": "https://github.com/necolas/react-native-web/tree/master/packages/dom-event-testing-library"
}
