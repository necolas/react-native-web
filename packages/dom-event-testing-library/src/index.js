/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

import { buttonType, buttonsType } from './constants';
import * as domEvents from './domEvents';
import * as domEventSequences from './domEventSequences';
import { hasPointerEvent, setPointerEvent, platform } from './domEnvironment';
import { describeWithPointerEvent, testWithPointerType } from './testHelpers';

const createEventTarget = (node) => ({
  node,
  /**
   * Simple events abstraction.
   */
  blur(payload) {
    node.dispatchEvent(domEvents.blur(payload));
  },
  click(payload) {
    node.dispatchEvent(domEvents.click(payload));
  },
  contextmenu(payload) {
    domEventSequences.contextmenu(node, payload);
  },
  error() {
    node.dispatchEvent(domEvents.error());
  },
  focus(payload) {
    domEventSequences.focus(node, payload);
    try {
      node.focus();
    } catch (e) {}
  },
  keydown(payload) {
    node.dispatchEvent(domEvents.keydown(payload));
  },
  keyup(payload) {
    node.dispatchEvent(domEvents.keyup(payload));
  },
  load(payload) {
    node.dispatchEvent(domEvents.load(payload));
  },
  /**
   * PointerEvent abstraction.
   * Dispatches the expected sequence of PointerEvents, MouseEvents, and
   * TouchEvents for a given environment.
   */
  // node no longer receives events for the pointer
  pointercancel(payload) {
    domEventSequences.pointercancel(node, payload);
  },
  // node dispatches down events
  pointerdown(payload) {
    domEventSequences.pointerdown(node, payload);
  },
  // node dispatches move events (pointer is not down)
  pointerhover(payload) {
    domEventSequences.pointerhover(node, payload);
  },
  // node dispatches move events (pointer is down)
  pointermove(payload) {
    domEventSequences.pointermove(node, payload);
  },
  // node dispatches enter & over events
  pointerover(payload) {
    domEventSequences.pointerover(node, payload);
  },
  // node dispatches exit & leave events
  pointerout(payload) {
    domEventSequences.pointerout(node, payload);
  },
  // node dispatches up events
  pointerup(payload) {
    domEventSequences.pointerup(node, payload);
  },
  scroll(payload) {
    node.dispatchEvent(domEvents.scroll(payload));
  },
  select(payload) {
    node.dispatchEvent(domEvents.select(payload));
  },
  // selectionchange is only dispatched on 'document'
  selectionchange(payload) {
    document.dispatchEvent(domEvents.selectionchange(payload));
  },
  /**
   * Gesture abstractions.
   * Helpers for event sequences expected in a gesture.
   * target.tap({ pointerType: 'touch' })
   */
  tap(payload) {
    domEventSequences.pointerdown(payload);
    domEventSequences.pointerup(payload);
  },
  virtualclick(payload) {
    node.dispatchEvent(domEvents.virtualclick(payload));
  },
  /**
   * Utilities
   */
  setBoundingClientRect({ x, y, width, height }) {
    node.getBoundingClientRect = function () {
      return {
        width,
        height,
        left: x,
        right: x + width,
        top: y,
        bottom: y + height
      };
    };
  }
});

const clearPointers = domEventSequences.clearPointers;

export {
  buttonType,
  buttonsType,
  clearPointers,
  createEventTarget,
  describeWithPointerEvent,
  platform,
  hasPointerEvent,
  setPointerEvent,
  testWithPointerType
};
