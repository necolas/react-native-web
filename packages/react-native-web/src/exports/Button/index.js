/**
 * Copyright (c) Nicolas Gallagher.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import * as React from 'react';
import StyleSheet from '../StyleSheet';
import TouchableOpacity from '../TouchableOpacity';
import Text from '../Text';

type ButtonProps = {|
  accessibilityLabel?: ?string,
  color?: ?string,
  disabled?: boolean,
  onPress?: ?(e: any) => void,
  testID?: ?string,
  title: string
|};

const Button = React.forwardRef<ButtonProps, *>((props, forwardedRef) => {
  const { accessibilityLabel, color, disabled, onPress, testID, title } = props;

  return (
    <TouchableOpacity
      accessibilityLabel={accessibilityLabel}
      accessibilityRole="button"
      disabled={disabled}
      focusable={!disabled}
      onPress={onPress}
      ref={forwardedRef}
      style={[
        styles.button,
        color && { backgroundColor: color },
        disabled && styles.buttonDisabled
      ]}
      testID={testID}
    >
      <Text style={[styles.text, disabled && styles.textDisabled]}>{title}</Text>
    </TouchableOpacity>
  );
});

Button.displayName = 'Button';

const styles = StyleSheet.create({
  button: {
    backgroundColor: '#2196F3',
    borderRadius: 2
  },
  text: {
    color: '#fff',
    fontWeight: '500',
    padding: 8,
    textAlign: 'center',
    textTransform: 'uppercase'
  },
  buttonDisabled: {
    backgroundColor: '#dfdfdf'
  },
  textDisabled: {
    color: '#a1a1a1'
  }
});

export type { ButtonProps };

export default Button;
