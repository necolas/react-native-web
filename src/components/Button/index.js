import ColorPropType from '../../propTypes/ColorPropType';
import React, { Component, PropTypes } from 'react';
import StyleSheet from '../../apis/StyleSheet';
import TouchableOpacity from '../Touchable/TouchableOpacity';
import Text from '../Text';

class Button extends Component {
  static propTypes = {
    accessibilityLabel: PropTypes.string,
    color: ColorPropType,
    disabled: PropTypes.bool,
    onPress: PropTypes.func.isRequired,
    title: PropTypes.string.isRequired
  };

  render() {
    const {
      accessibilityLabel,
      color,
      disabled,
      onPress,
      title
    } = this.props;

    return (
      <TouchableOpacity
        accessibilityLabel={accessibilityLabel}
        accessibilityRole={'button'}
        disabled={disabled}
        onPress={onPress}
        style={[
          styles.button,
          color && { backgroundColor: color },
          disabled && styles.buttonDisabled
        ]}>
        <Text style={[
          styles.text,
          disabled && styles.textDisabled
        ]}>
          {title}
        </Text>
      </TouchableOpacity>
    );
  }
}

const styles = StyleSheet.create({
  button: {
    backgroundColor: '#2196F3',
    borderRadius: 2
  },
  text: {
    textAlign: 'center',
    color: '#fff',
    padding: 8,
    fontWeight: '500'
  },
  buttonDisabled: {
    backgroundColor: '#dfdfdf'
  },
  textDisabled: {
    color: '#a1a1a1'
  }
});

module.exports = Button;
