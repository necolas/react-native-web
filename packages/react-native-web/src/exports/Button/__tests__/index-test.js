import Button from '..';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { createEventTarget } from 'dom-event-testing-library';
import { render } from '@testing-library/react';

describe('components/Button', () => {
  test('prop "accessibilityLabel"', () => {
    const { container } = render(
      <Button accessibilityLabel="accessibility label" title="" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('prop "color"', () => {
    const color = 'rgb(0, 0, 255)';
    const { container } = render(<Button color={color} title="" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('prop "disabled"', () => {
    const { container } = render(<Button disabled={true} title="" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('prop "onPress"', () => {
    const onPress = jest.fn();
    const ref = React.createRef();
    act(() => {
      render(<Button onPress={onPress} ref={ref} title="" />);
    });
    const target = createEventTarget(ref.current);
    act(() => {
      target.pointerdown({ button: 0 });
      target.pointerup({ button: 0 });
    });
    expect(onPress).toBeCalled();
  });

  test('prop "testID"', () => {
    const { container } = render(<Button testID="123" title="" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('prop "title"', () => {
    const { container } = render(<Button title="Click me" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
