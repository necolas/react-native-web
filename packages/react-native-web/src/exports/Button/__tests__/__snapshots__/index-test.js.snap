// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Button prop "accessibilityLabel" 1`] = `
<div
  aria-label="accessibility label"
  class="css-view-175oi2r r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-cursor-1loqt21 r-touchAction-1otgn73 r-backgroundColor-14sbq61 r-borderRadius-1jkafct"
  role="button"
  style="transition-duration: 0s;"
  tabindex="0"
>
  <div
    class="css-text-1rynq56 r-color-jwli3a r-fontWeight-majxgm r-padding-edyy15 r-textAlign-q4m81j r-textTransform-tsynxw"
    dir="auto"
  />
</div>
`;

exports[`components/Button prop "color" 1`] = `
<div
  class="css-view-175oi2r r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-cursor-1loqt21 r-touchAction-1otgn73 r-borderRadius-1jkafct"
  role="button"
  style="background-color: rgb(0, 0, 255); transition-duration: 0s;"
  tabindex="0"
>
  <div
    class="css-text-1rynq56 r-color-jwli3a r-fontWeight-majxgm r-padding-edyy15 r-textAlign-q4m81j r-textTransform-tsynxw"
    dir="auto"
  />
</div>
`;

exports[`components/Button prop "disabled" 1`] = `
<div
  aria-disabled="true"
  class="css-view-175oi2r r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-borderRadius-1jkafct r-backgroundColor-11mpjr4 r-pointerEvents-633pao"
  role="button"
  style="transition-duration: 0s;"
  tabindex="-1"
>
  <div
    class="css-text-1rynq56 r-fontWeight-majxgm r-padding-edyy15 r-textAlign-q4m81j r-textTransform-tsynxw r-color-c68hjy"
    dir="auto"
  />
</div>
`;

exports[`components/Button prop "testID" 1`] = `
<div
  class="css-view-175oi2r r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-cursor-1loqt21 r-touchAction-1otgn73 r-backgroundColor-14sbq61 r-borderRadius-1jkafct"
  data-testid="123"
  role="button"
  style="transition-duration: 0s;"
  tabindex="0"
>
  <div
    class="css-text-1rynq56 r-color-jwli3a r-fontWeight-majxgm r-padding-edyy15 r-textAlign-q4m81j r-textTransform-tsynxw"
    dir="auto"
  />
</div>
`;

exports[`components/Button prop "title" 1`] = `
<div
  class="css-view-175oi2r r-transitionProperty-1i6wzkk r-userSelect-lrvibr r-cursor-1loqt21 r-touchAction-1otgn73 r-backgroundColor-14sbq61 r-borderRadius-1jkafct"
  role="button"
  style="transition-duration: 0s;"
  tabindex="0"
>
  <div
    class="css-text-1rynq56 r-color-jwli3a r-fontWeight-majxgm r-padding-edyy15 r-textAlign-q4m81j r-textTransform-tsynxw"
    dir="auto"
  >
    Click me
  </div>
</div>
`;
